tools/websoc-scraper-v2/index.ts

```diff
@@ -0,0 +1,453 @@
+   */
+  ge: Record<string, WebsocAPIResponse>;
+};
+
+/**
+ * An instructor object that can be inserted directly into Prisma.
+ */
+type ProcessedInstructor = {
+  year: string;
+  quarter: Quarter;
+  sectionCode: number;
+  timestamp: Date;
+  name: string;
+};
+
+/**
+ * A meeting object that can be inserted directly into Prisma.
+ */
+type ProcessedMeeting = {
+  year: string;
+  quarter: Quarter;
+  sectionCode: number;
+  timestamp: Date;
+  days: string[];
+  buildings: string[];
+  startTime: number;
+  endTime: number;
+};
+
+/**
+ * A section object containing additional metadata.
+ */
+type ProcessedSection = {
+  /**
+   * Contains the instructors and meetings associated with this section.
+   */
+  meta: {
+    instructors: ProcessedInstructor[];
+    meetings: ProcessedMeeting[];
+  };
+  /**
+   * The section object that can be inserted directly into Prisma.
+   */
+  data: {
+    year: string;
+    quarter: Quarter;
+    sectionCode: number;
+    timestamp: Date;
+    geCategories: GE[];
+    department: string;
+    courseNumber: string;
+    courseNumeric: number;
+    courseTitle: string;
+    sectionType: (typeof sectionTypes)[number];
+    units: string;
+    maxCapacity: number;
+    sectionFull: boolean;
+    waitlistFull: boolean;
+    overEnrolled: boolean;
+    cancelled: boolean;
+    data: object;
+  };
+};
+
+// The duration to sleep between scraping runs, or to wait before querying the
+// database for terms to scrape on demand if there were none in the last query.
+const SLEEP_DURATION = 5 * 60 * 1000; // 5 minutes
+
+const prisma = new PrismaClient();
+
+const logger = createLogger({
+  level: "info",
+  format: format.combine(
+    process.env.NODE_ENV === "development"
+      ? format.colorize({ all: true })
+      : format.uncolorize(),
+    format.timestamp(),
+    format.printf((info) => `${info.timestamp} [${info.level}] ${info.message}`)
+  ),
+  transports: [new transports.Console()],
+  exitOnError: false,
+});
+
+/**
+ * Sleep for the given number of milliseconds.
+ * @param duration Duration in ms.
+ */
+const sleep = async (duration: number) =>
+  new Promise((resolve) => setTimeout(resolve, duration));
+
+/**
+ * Get all terms that are to be scraped on a daily basis.
+ * @param date The current date.
+ */
+async function getTermsToScrape(date: Date) {
+  const quarterDates = Object.assign(
+    {},
+    ...(await Promise.all(
+      Array.from(Array(3).keys()).map((x) =>
+        getTermDateData((x + date.getFullYear() - 1).toString())
+      )
+    ))
+  );
+  return Object.fromEntries(
+    (await getTerms())
+      .map((term) => term.shortName)
+      .filter((term) => Object.keys(quarterDates).includes(term))
+      .filter((term) => date <= quarterDates[term].finalsStart)
+      .map(
+        (term) =>
+          [term, { year: term.split(" ")[0], quarter: term.split(" ")[1] }] as [
+            string,
+            Term
+          ]
+      )
+  );
```

Comment:
```suggestion
  type TermEntry = [string, Term]

  const terms = await getTerms()
  const termEntries = terms
    .map((term) => term.shortName)
    .filter((term) => Object.keys(quarterDates).includes(term))
    .filter((term) => date <= quarterDates[term].finalsStart)
    .map((term) => {
      const termEntry: TermEntry = [term, { year: term.split(" ")[0], quarter: term.split(" ")[1] }]
      return termEntry
    })
  return Object.fromEntries(termEntries);
```
This sequence is more verbose and uses additional variables, but I think it really aids with logical clarity. It's also easier to view type definitions of each step to see how data is transformed to produce the output.